### PR TITLE
feat(b-sidebar): new custom component `<b-sidebar>` (closes #3324, #3210, #1702)

### LINKED ISSUE #1702: SideNav a.k.a. Drawer component
Hi, I would like to suggest a drawer component in new release. It it the only one component that is missing for me in BootstrapVue.

### LINKED ISSUE #3210: Doc: Please provide an example implementation of the classical left drawer
Please provide an example implementation of the classical left drawer.

I mean, a left column with items, and a button on navbar to hide/show it.

Probably it's all easy to do, but for my, a newbie with this package, it's a completely caos to understand where/what to add; and problably there is a ready-to-use component and I am not seeing it...

thanks for this excellent work, and for your help in advance

### LINKED ISSUE #3324: Is there a sidebar component?
Is there a sidebar component or what would be the recommended way to create a side navbar? that can also show/hide?

### PR DESCRIPTION
**Pull request by tmorehouse**
### Describe the PR

New `<b-sidebar>` component, for creating fixed viewport (left or right) slide out drawers.

Users can place any arbitrary content they wish in the sidebars.

Very basic component, that leaves the majority of styling and content layout up to the user.

Roughly based on the Bootstrap `offcanvas` example.

Closes #3324
Closes #3210
Closes #1702

To Do:

- [x] make header with close button optional
- [x] optional `footer` slot
- [x] accessibility docs
- [x] `lazy` prop docs
- [x] unit testing
- [x] SASS variables
- [x] other stuff (TBD)
  - [x] close on ESC keypress (opt-out)
  - [x] close on route change (opt-out) uses fullPath including query and hash
  - [x] add optional `lazy` content rendering
  - [x] handle tooltips/popover inside sidebars
  -  ~close on clickout/focusout? (opt-in or out ? as it may close on things like interactive popovers)~ leaving out for now.
  -  ~add docs page for `v-b-toggle` directive~ (make a separate PR)

### PR checklist

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Enhancement
- [ ] ARIA accessibility
- [ ] Documentation update
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** (check one)

- [x] No
- [ ] Yes (please describe)

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, **not** the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (i.e. `[...] (fixes #xxx[,#xxx])`, where "xxx" is the issue number)
- [x] It should address only one issue or feature. If adding multiple features or fixing a bug and adding a new feature, break them into separate PRs if at all possible.
- [x] The title should follow the [**Conventional Commits**](https://www.conventionalcommits.org/) naming convention (i.e. `fix(alert): not alerting during SSR render`, `docs(badge): update pill examples`, `chore(docs): fix typo in README`, etc). **This is very important, as the `CHANGELOG` is generated from these messages.**

**If new features/enhancement/fixes are added or changed:**

- [x] Includes documentation updates (including updating the component's `package.json` for slot and event changes)
- [x] Includes any needed TypeScript declaration file updates
- [ ] New/updated tests are included and passing (if required)
- [x] Existing test suites are passing
- [ ] CodeCov for patch has met target
- [x] The changes have not impacted the functionality of other components or directives
- [x] ARIA Accessibility has been taken into consideration (Does it affect screen reader users or keyboard only users? Clickable items should be in the tab index, etc.)

**If adding a new feature, or changing the functionality of an existing feature, the PR's
description above includes:**

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: src/components/sidebar/sidebar.js [hunk 1]
```diff
@@ -24,23 +31,23 @@ export const BSidebar = /*#__PURE__*/ Vue.extend({
     },
     right: {
       type: Boolean,
-      default: false
+      default: () => getComponentConfig(NAME, 'right')
```
Comment (tmorehouse) [inline src/components/sidebar/sidebar.js]: We don't allow boolean only props to be in the config.

So this should be hard coded to `false`

File: src/components/sidebar/sidebar.js [hunk 2]
```diff
@@ -55,21 +62,26 @@ export const BSidebar = /*#__PURE__*/ Vue.extend({
...
       // default: null
     },
+    contentClass: {
+      type: [String, Array, Object]
+      // default: null
+    },
     noSlide: {
       type: Boolean,
-      default: false
+      default: () => getComponentConfig(NAME, 'noSlide')
```
Comment (tmorehouse) [inline src/components/sidebar/sidebar.js]: We don't allow boolean only props to be in the config.

So this should be hard coded to `false`

File: src/components/sidebar/sidebar.js [hunk 3]
```diff
@@ -194,18 +214,39 @@ export const BSidebar = /*#__PURE__*/ Vue.extend({
...
+      {
+        staticClass: `${CLASS_NAME}-content`,
+        class: this.contentClass
+      },
+      this.normalizeSlot('default', scope)
+    )
+
+    let $body = h()
+    if (!(this.lazy && !this.localShow)) {
+      $body = h('div', { staticClass: `${CLASS_NAME}-body` }, [$header, $content])
```
Comment (tmorehouse) [inline src/components/sidebar/sidebar.js]: Header should not be part of body... it should be outside of it.

As the header needs to be there for accessibility reason when hidden (as it acts as the accessible label for the sidebar)

File: src/components/sidebar/sidebar.js (hunk 1 above)
Comment (tmorehouse) [reply src/components/sidebar/sidebar.js]: fixed

File: src/components/sidebar/sidebar.js (hunk 2 above)
Comment (tmorehouse) [reply src/components/sidebar/sidebar.js]: fixed

**Comment by jensgeffken** [comment]
is this already implemented ? i am looking for a sidebar offcanvas menu in bootstrap-vue. Thanks

**Comment by tmorehouse** [comment]
@jensgeffken it has been available since BootstrapVue v2.10.0

**Comment by bbugh** [comment]
Is bootstrap-vue breaking from feature parity Bootstrap, or will this sidebar be updated to use the [offcanvas component](https://deploy-preview-29017--twbs-bootstrap.netlify.app/docs/5.0/components/offcanvas/) from Bootstrap 5 (if it is merged)?

**Comment by jackmu95** [comment]
@bbugh It will be updated to match Bootstrap v5's Offcanvas component in the future.

**Comment by mangrove108** [comment]
does anyone knows how many breaking changes are coming. in bootstrap 5 ??? 😟

**Pull request merged**